### DISCUSSION TITLE
Low performance of igl::copyleft::cgal::mesh_boolean ?

### DISCUSSION THREAD

**Discussion by AntonChernobrovkin** [libigl/libigl#2367] [General]
I measured performance of igl::copyleft::cgal::mesh_boolean, and boolean operation in cgal directly. I expected the similar performance, but my measures showed that libigl slower than cgal in 3-6 times for all cases.
For same operation it took 7.5s for cgal and 40s for libigl. On another pair: 1.6.s for cgal, 5.2s for libigl.
Is this expected?

Notes:
-OS: Windows, compiled with Visual Studio for both cases.
-Release, not debug.
-Reproduced for all operations: union, intersection, diff, xor.

Libigl code:
```
Eigen::MatrixXd V1, V2, V3;
Eigen::MatrixXi F1, F2, F3;

igl::readOBJ("mesh_a.obj", V1, F1);
igl::readOBJ("mesh_b.obj", V2, F2);

auto startTime = std::chrono::high_resolution_clock::now();
igl::copyleft::cgal::mesh_boolean(V1, F1, V2, F2, igl::MeshBooleanType::MESH_BOOLEAN_TYPE_UNION, V3, F3);
std::chrono::duration<double> seconds = std::chrono::high_resolution_clock::now() - startTime;
std::cout << seconds.count() << " sec \n";
```

cgal code:
```
namespace PMP = CGAL::Polygon_mesh_processing;

PMP::IO::read_polygon_mesh("mesh_a.obj", mesh1)
PMP::IO::read_polygon_mesh("mesh_b.obj", mesh2))
valid_union = PMP::corefine_and_compute_union(mesh1, mesh2, out);

auto startTime = std::chrono::high_resolution_clock::now();
valid_union = PMP::corefine_and_compute_union(mesh1, mesh2, out);
std::chrono::duration<double> seconds = std::chrono::high_resolution_clock::now() - startTime;
std::cout << seconds.count() << " sec \n";
```

**Comment by alecjacobson** [comment]
Could you please share your inputs?

**Reply by AntonChernobrovkin** [reply]
Here it is:
https://drive.google.com/file/d/1hUqbd1AGlWAO1QRT3QOjW_Kd5Zd18ulj/view?usp=sharing
https://drive.google.com/file/d/1N8GMCaYPqEFZ-SAepC2gX3JJ7S_D2Dih/view?usp=sharing

Basicly its are the same meshes with different spatial positions. I believe it doesn't affect the performance.
Here I measured 7.5s for cgal and 40s for libigl.

**Reply by alecjacobson** [reply]
OK, I tried mesh_b union mesh_b+[50 0 0] and I also got 40s. CGAL's booleans have improved _a lot_ since the time of publishing our boolean method. I'm guessing if you meet their more strict input preconditions you should use those.

**Reply by AntonChernobrovkin** [reply]
Thank you for clarification!
